census_view.cpp

```diff
@@ -932,6 +935,969 @@ inline std::vector<std::string> const& CensusViewDataViewModel::all_headers() co
     return view_.case_parms()[0].member_names();
 }
 
+namespace
+{
+
+// Declare the functions here, but implement after the CensusViewGridTable
+// declaration to avoid the "use of undefined type" error.
+
+// Get the cell value from the table.
+
+any_member<Input> const& cell_at(wxGridTableBase const& table, int row, int col);
+
+// Get the raw value from the table.
+
+Input const& row_at(wxGridTableBase const& table, int row);
+
+// Get the column name from the table.
+
+std::string const& col_name(wxGridTableBase const& table, int col);
+
+// class RangeDoubleEditor
+
+class RangeDoubleEditor
+    :public wxGridCellTextEditor
+{
+  public:
+    RangeDoubleEditor() = default;
+
+    void BeginEdit(int row, int col, wxGrid* grid) override
+    {
+        auto const& value = cell_at(*grid->GetTable(), row, col);
+        auto const* as_range = member_cast<tn_range_base>(value);
+        LMI_ASSERT(typeid(double) == as_range->value_type());
+
+        wxFloatingPointValidator<double> val;
+        val.SetRange(as_range->universal_minimum(), as_range->universal_maximum());
+        SetValidator(val);
+
+        wxGridCellTextEditor::BeginEdit(row, col, grid);
+    }
+};
+
+// class RangeDateEditor
+
+class RangeDateEditor
+    :public wxGridCellDateEditor
+{
+  public:
+    RangeDateEditor() = default;
+
+    void BeginEdit(int row, int col, wxGrid* grid) override
+    {
+        auto const& value = cell_at(*grid->GetTable(), row, col);
+        auto const* as_range = member_cast<tn_range_base>(value);
+        LMI_ASSERT(typeid(calendar_date) == as_range->value_type());
+
+        DatePicker()->SetValue(ConvertDateToWx(value_cast<calendar_date>(value)));
+
+        auto const min_value = as_range->universal_minimum();
+        auto const max_value = as_range->universal_maximum();
+
+        DatePicker()->SetRange
+            (ConvertDateToWx(jdn_t(bourn_cast<int>(min_value)))
+            ,ConvertDateToWx(jdn_t(bourn_cast<int>(max_value)))
+            );
+
+        DatePicker()->SetFocus();
+    }
+};
+
+// class DatumSequenceEditor
+
+class DatumSequenceEditorEvtHandler
+    :public wxEvtHandler
+{
+  public:
+    // wxIMPLEMENT_DYNAMIC_CLASS requires the default constructor
+    // so add the default value for the entry parameter.
+    explicit DatumSequenceEditorEvtHandler(InputSequenceEntry* entry = nullptr)
+        :entry_(entry)
+    {
+    }
+
+    void UponChar(wxKeyEvent& event)
+    {
+        switch(event.GetKeyCode())
+            {
+            case WXK_RETURN:
+            case WXK_NUMPAD_ENTER:
+                if(!wxGetKeyState(WXK_ALT))
+                    {
+                    event.Skip();
+                    return;
+                    }
+
+                // Just handle the event to not allow other handlers
+                // process it. The editor window opened in OnKeyDown.
+                break;
+            default:
+                event.Skip();
+                break;
+            }
+    }
```

Comment:
I'm removing this because it doesn't seem to be necessary -- I don't see any difference in behaviour with and without this handler. Please let me know if I'm missing anything here (otherwise, please resolve this conversation).